### PR TITLE
Return commit hash to tendermint

### LINKED ISSUE #1829: Send the hash to tendermint
_description will be added soon_

### PR DESCRIPTION
**Pull request by kansi**
Fixes #1829

### CODE REVIEW

File: bigchaindb/tendermint/core.py [hunk 1]
```diff
@@ -46,4 +50,11 @@ def deliver_tx(self, raw_transaction):
             return Result.error(log='Invalid transaction')
         else:
             self.bigchaindb.store_transaction(transaction)
+            self.block_txn_ids.append(transaction.to_dict()['id'])
             return Result.ok()
+
+    def commit(self):
+        """ Return merkle root of the transactions included in the
+        block"""
+        merkle_root = calculate_hash(self.block_txn_ids)
```
Comment (kansi) [inline bigchaindb/tendermint/core.py]: Should move this to `end_block` because `commit` is blocking

File: bigchaindb/tendermint/utils.py [hunk 2]
```diff
@@ -12,3 +13,11 @@ def decode_transaction(raw):
     """Decode a transaction from Base64 to a dict."""
 
     return json.loads(raw.decode('utf8'))
+
+
+def calculate_hash(key_list):
```
Comment (ttmc) [inline bigchaindb/tendermint/utils.py]: Does Tendermint expect a SHA-256 hash (i.e. a SHA-2 hash), or can we give it a SHA3-256 hash instead? BigchainDB usually uses SHA3-256 hashes. For example, the ID of a BigchainDB transaction is a SHA3-256 hash (represented as a 64-character hex string. 256 bits = 32 bytes and each byte can be represented by two hex characters).

Comment (kansi) [reply bigchaindb/tendermint/utils.py]: Tendermint doesn't seem to specify any such contraint, so I have moved to SHA3-256.

File: bigchaindb/backend/localmongodb/query.py [hunk 3]
```diff
@@ -39,3 +40,20 @@ def get_spent(conn, transaction_id, output):
...
+
+@register_query(LocalMongoDBConnection)
+def get_latest_block(conn):
+    return conn.run(
+        conn.collection('blocks')
+        .find_one(sort=[('height', DESCENDING)]))
+
+
+@register_query(LocalMongoDBConnection)
+def create_new_block(conn, block):
```
Comment (vrde) [inline bigchaindb/backend/localmongodb/query.py]: We have other methods called `store_<obj>`, like `store_transaction` or `store_asset`. Following this pattern, we can call this method `store_block`, what do you think?

File: bigchaindb/tendermint/core.py [hunk 4]
```diff
@@ -1,11 +1,12 @@
 """This module contains all the goodness to integrate BigchainDB
 with Tendermint."""
 
-
 from abci import BaseApplication, Result
+from abci.types_pb2 import ResponseEndBlock, ResponseInfo
```
Comment (vrde) [inline bigchaindb/tendermint/core.py]: You can import everything from `abci`. As you can see from their [`abci/__init__.py`](https://github.com/davebryson/py-abci/blob/master/abci/__init__.py) they export all the useful classes.

File: bigchaindb/tendermint/core.py [hunk 5]
```diff
@@ -19,6 +20,29 @@ def __init__(self, bigchaindb=None):
...
             bigchaindb = BigchainDB()
         self.bigchaindb = bigchaindb
+        self.block_txn_ids = []
+        self.block_txn_hash = ''
+        self.validators = None
+        self.new_height = None
+
+    def init_chain(self, validators):
+        """Initialize chain with block of height 0"""
+        print("\ninit_chain::\n")
```
Comment (vrde) [inline bigchaindb/tendermint/core.py]: I'd use `log`. Or you can even remove the `print` for now, and we address logging later.

File: bigchaindb/backend/localmongodb/query.py (hunk 3 above)
Comment (kansi) [reply bigchaindb/backend/localmongodb/query.py]: sure

File: bigchaindb/tendermint/core.py (hunk 5 above)
Comment (kansi) [reply bigchaindb/tendermint/core.py]: Sure, will remove it. Was using it for testing locally.

File: bigchaindb/tendermint/core.py (hunk 4 above)
Comment (kansi) [reply bigchaindb/tendermint/core.py]: I can see`ResponseInfo` is exported but `ResponseEndBlock` is not exported there.

**Pull request merged**